### DISCUSSION TITLE
Variable undefined

### DISCUSSION THREAD

**Discussion by gerukin** [AnWeber/vscode-httpyac#47] [General]
Hi, I'm new to `httpyac`. I currently use `rest-client` and `Postman`. The extension and runner look promising, congrats!, but somehow I cannot make anything work.

I'm using the latest general availability vscode (Mac) and have installed both the cli and the extension. I opened a new folder with just this one file:
```
@url = http://www.google.com

### Test
{{url}}
```

Clicking `send` gets me this:

```
ERROR: js: exports.$result = (url);
ERROR: ReferenceError - url is not defined
ReferenceError: url is not defined
	at Object.userJS (/Users/gerukin/dev/http-test/test.http:4:86)
	at Object.<anonymous> (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:37501)
	at Generator.next (<anonymous>)
	at /Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:36506
	at new Promise (<anonymous>)
	at s (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:36251)
	at Object.g (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:36785)
	at t.JavascriptVariableReplacer.<anonymous> (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:143455)
	at Generator.next (<anonymous>)
	at /Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:142841
	at new Promise (<anonymous>)
	at r (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:142586)
	at t.JavascriptVariableReplacer.replace (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:143089)
	at Object.<anonymous> (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:46841)
	at Generator.next (<anonymous>)
	at s (/Users/gerukin/.vscode/extensions/anweber.vscode-httpyac-2.18.0/dist/extension.js:587:46333)
	at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

This works though:
```
### Test
@url = http://www.google.com
{{url}}
```

But my understanding is that those variables are in a global scope and should be available to all requests. What am I doing wrong?

**Comment by AnWeber** [comment]
Your understanding is correct.  Spontaneously, I do not see the error. I'll get back to you when I have more time.

**Comment by AnWeber** [comment]
also outline looks valid
![image](https://user-images.githubusercontent.com/13434739/124612838-f308a480-de72-11eb-83e8-f86a8b87a2ca.png)

**Reply by gerukin** [reply]
Thank you!

My outline looks slightly different from yours:
<img width="250" alt="Screen Shot 2021-07-06 at 23 09 58" src="https://user-images.githubusercontent.com/36127103/124614594-5ca6a400-deaf-11eb-8c10-07991e720962.png">

**Comment by AnWeber** [comment]
I tried some modifications, but got the same error. I think, I broke something with the global variables. I mostly use env files, so it is a not so tested feature.

**Reply by gerukin** [reply]
Ah, well in a way I'm happy you are able to reproduce it. 😁 This extension must have been a lot of work and many moving parts. Thanks for looking into it! I suppose this can be converted to an issue then.

**Reply by AnWeber** [reply]
The cause was a stupid error with an if. I should really start writing unit tests:-) I released a new version (2.18.1, npm and vscode marketplace)
Initially it was a fun coding project because of Lockdown and Covid-19, but I really enjoy it and have learned/seen other things as well. That's why today it's about 80-100 hours (or more). Next I want to add a more rich interface for extensions. In case I lose the desire/ time, so that the project is not a dead end. I thought about a concept like vue-cli plugins, that would be 16 hours more:-)

**Reply by gerukin** [reply]
Thanks, that was fast. It works fine now. 🎉

Honestly it's a pretty good extension and the runner is really useful. Trying to see if I can ditch POSTMAN entirely as well. Truth be told, my main concern at this point is stability of the runner and finding a way to make everything work for both vscode and PHPstorm users (which I see you've done a ton of work to make easier to attempt).

**Comment by AnWeber** [comment]
PhpStorm/ Intellij Idea Syntax should work pretty well. I wanted to implement a superset of Intellij and vscode rest-client.  If you find any discrepancies, it would be nice to tell me, since I actually want to be completely compatible. I would build myself a template in PhpStorm once, and only then really use my extension. Good links are:
https://www.jetbrains.com/help/idea/http-client-in-product-code-editor.html
https://github.com/JetBrains/http-request-in-editor-spec/blob/master/spec.md

Unfortunately, I agree with you about stability. And it would be just as much my concern from the outside. I'm currently trying to conceal the weak point with quick bug fixes. But there is really no way around creating unit tests. This is currently a question of time, as I have two lively children. But I am working on it.

**Reply by gerukin** [reply]
I hear you. Life goes so fast. Thank you for the part of yours you are putting into this extension. I'm sure it is saving a lot of people a lot of time. 🙇‍♂️

I don't use PHPstorm myself so I can only test the rest-client compatibility, but I'll be sharing test scripts with coworkers who do. So if we run into issues I'll let you know, thank you!

That being said, no amount of praise should ever go without some counterbalancing bug report.. so here is something else (related) that I've noticed. I'm struggling with imports. I simply moved the `@url` to a `config.http` file in the same folder. And now I have this:
```
# @import ./config.http

### Test
{{url}}
```

It gives me the same error as before. Looking at your docs, I was under the impression all variables from an imported file would be available in the global scope. And `@ref` could be used for named requests from the imported file. Am I missing something?

**Reply by AnWeber** [reply]
:-) damn, it was never intended that way. But yes, that's how you can understand it and that's how it would work if you import a request using @ref. So in principle an issue.
You can work around it currently, if you assign a name and import this.

```
@url = http://www.google.com
# @name global
```

```
# @import ./config.http
# @ref global

### Test
{{url}} 
```
The nice thing is, you can basically turn it into a feature that solves the idea of Huachao/vscode-restclient#845. Different as described, but the same idea. I will have a look at it in the next few days.

By the way, this is not conform to PhpStorm, here the variables are stored in `http-client.env.json` files and it is also not implemented in rest-client.

**Reply by gerukin** [reply]
Yes indeed, since I'm trying to stick to PHPstorm compatibility for at least some projects I won't be using it there. I'm starting to understand the limitations a the various IDEs and tools a bit better here. Thank you.

Speaking of limitations.. I tried to nest and organize my code in folders but it looks like this is not a supported structure. For example:
```
|__ http-client.env.json
|__ system
   |__ health.http
```
Doesn't allow me to use the environment variables from `http-client.env.json` in `health.http`. Maybe this is also the way it works in IntelliJ. Looking at [their docs](https://www.jetbrains.com/help/idea/exploring-http-syntax.html#using_request_vars) it's not clear to me.

I didn't like that `rest-client` forces env variables into vscode's settings rather than allow us to put them in the project dir, and I do like that httpyac supports both `.env` (my preferred format, allowing comments and all) and IntelliJ style json files (I'll survive)... but it feels like httpyac has more flexibility with `.env` files (since it looks like we can define an `env` directory and a root project directory with just a `package.json` file).

I'm sorry to bother you with this but what is the best practice for a multi dir structure? Preferably compatible with IntelliJ. 😎😇

**Comment by AnWeber** [comment]
my project structure is following (no intellij support, because of .env), but I have deliberately few restrictions on the folder structure in my extensions.
```
|__ env
   |__ .env
   |__ dev.env
   |__ prod.env
|__ src
   |__ health.http
   |__ app1
       |__ ...
   |__ app2
       |__ ...
```
The intellij file should be located in the root of the project, but can be moved to any location using setting `httpyac.intellijDirname`.
It may be because I only monitor the files if they are already present when vscode starts. If they are created afterwards, you have to run `httpyac.reset` once or restart vscode. After that a watcher is registered. If it does not work after the reset. enable Log=trace and the extensions outputs all search folder.
The behavior of `rest-client` was not mine either, hence this deviation here. And how exactly it works in PhpStorm I'm not sure, but I think root should fit (see https://github.com/AnWeber/vscode-httpyac/issues/9#issuecomment-770788483).

**Reply by gerukin** [reply]
A restart did the trick thank you. :)

Now I will have to make sure this also works in PhpStorm. I too would prefer to use `.env` files though but that's not an option for cross IDE work.

Last thing.. I have a couple small suggestions I'd like to share (one for the runner and one for the IDE extension). Do you prefer using the discussion boards or opening issues?

**Reply by AnWeber** [reply]
Definitely issues. I want to link changes in code to the report. If I view other repos I will search in issues and just sometimes in discussion.
